src/expression.c

```diff
@@ -9809,6 +9810,69 @@ Expression *SliceExp::syntaxCopy()
     return se;
 }
 
+bool SliceExp::isOpDollar(VarExp* ve)
+{
+    return ve->var == this->lengthVar;
+    // VarDeclaration* decl = ve->var->isVarDeclaration();
+    // return decl && decl->ident == Id::dollar;
+}
+
+SliceExp::Boundness SliceExp::analyzeRelativeBound(Expression* e,
+                                                   dinteger_t* p, dinteger_t* q) // out arguments
+{
+    if (e->op == TOKint64 && e->toInteger() == 0)
+    {
+        e->warning("Avoiding boundscheck for 0");
+        return atStart;
+    }
+    else if (e->op == TOKvar && this->isOpDollar((VarExp*)e)) // TODO functionize?
+    {
+        e->warning("Avoiding boundscheck for $");
+        return atEnd;
+    }
+    else if (e->op == TOKdiv)
+    {
+        DivExp* div = (DivExp*)e;
+        if (div->e1->op == TOKvar && this->isOpDollar((VarExp*)div->e1) && // TODO functionize?
+            div->e2->op == TOKint64)
+        {
+            *q = div->e2->toInteger();
+            if (*q >= 1)
+            {
+                e->warning("Avoiding boundscheck for $/%ld", *q);
+                return inside;
+            }
+        }
+        else if (div->e1->op == TOKmul &&
+                 div->e2->op == TOKint64)
+        {
+            MulExp* mul = (MulExp*)div->e1;
+            *q = div->e2->toInteger();
+            if (mul->e1->op == TOKvar && this->isOpDollar((VarExp*)mul->e1) && // TODO functionize?
+                mul->e2->op == TOKint64)
+            {
+                *p = mul->e2->toInteger();
+                if (*p <= *q)
+                {
+                    e->warning("Avoiding boundscheck for $*%ld/%ld", *p, *q);
+                    return inside;
+                }
+            }
+            else if (mul->e1->op == TOKint64 &&
+                     mul->e2->op == TOKvar && this->isOpDollar((VarExp*)mul->e2)) // TODO functionize?
+            {
+                *p = mul->e1->toInteger();
+                if (*p <= *q)
+                {
+                    e->warning("Avoiding boundscheck for $*%ld/%ld", *p, *q);
```

Comment:
`%ld` is not the correct format spec for `dinteger_t` on all platforms.  Instead, use `%lld` and cast the argument to `(unsigned long long)`